ptrace+strace: Halve number of syscalls done by strace

BenWiederhake opened the pull request:
The ptrace mechanism is silly: It requires dozens if not hundreds of syscalls by the tracer for each syscall by the tracee. That ratio alone is cause for excessive slowdown. One reason for this high ratio is the weird dance required to gather the various hints regarding why the tracee has stopped.

This PR fixes the *other* reason for the high ratio: Memory used to be copied in 4-byte increments. Now, memory is copied buffer-to-buffer, in a single syscall. This means less syscall traffic for the Kernel, and a saner interface, especially for future applications (yes, I have one in mind) that use ptrace to copy more than just a struct or two.

before:
![Bildschirmfoto_2021-11-26_20-26-37](https://user-images.githubusercontent.com/2690845/143769676-5910415e-3e4a-4737-86d6-1a864ca3554e.png)
after:
![Bildschirmfoto_2021-11-27_23-58-32](https://user-images.githubusercontent.com/2690845/143769680-cc1ea7c5-279d-4c21-8231-e223008d0b2a.png)

Sadly, this does not significantly improve the runtime impact of strace/ptrace:
before:
![before_pr](https://user-images.githubusercontent.com/2690845/143769705-42d673be-5875-4d22-9789-8253345f2f10.png)
after:
![after_pr](https://user-images.githubusercontent.com/2690845/143769712-e32e58be-a1e9-4813-a4b5-205b5b6ba732.png)

This seems to be due to other factors:
![Bildschirmfoto_2021-11-28_14-15-40](https://user-images.githubusercontent.com/2690845/143769727-000f41d2-ed74-452d-b9fc-77b19710de97.png)
My wild guess is that somehow ptrace starting/stopping a process interferes with the scheduler, and causes a lot of contention on the tracee's address space spinlock. This would explain why reducing the number of `PT_PEEK`s only has marginal effect on the running time.

Nevertheless, improving the running time wasn't my goal here, so I just declare improving the running time to be out-of-scope for this PR.

creator1creeper1 commented on Userland/Libraries/LibC/sys/ptrace.cpp:
```diff
@@ -12,6 +12,15 @@ extern "C" {
 {
+    if (request == PT_PEEKBUF) {
+        // PT_PEEKBUF cannot easily be correctly used through this function signature:
+        // The amount of data to be compied is not available.
```
`copied` instead of `compied`

BenWiederhake commented:
Changes since last push:
- Fixed silly typo (thanks creator1creeper1!)
- Rebased

BertalanD commented on AK/Format.h:
```diff
@@ -624,6 +624,16 @@ struct Formatter<Error> : Formatter<FormatString> {
+    ErrorOr<void> format(FormatBuilder& builder, ErrorOr<T, ErrorType> const& error_or)
+    {
+        if (error_or.is_error())
+            return Formatter<FormatString>::format(builder, "{}", error_or.error());
```
This creates a temporary `StringBuilder`, which might allocate memory.
 Would something like the following work?
```diff
            return Formatter<ErrorType>::format(builder, error_or.error());
```

BenWiederhake replied:
Oh yeah, you're right!

BenWiederhake replied:
`Formatter<FormatString>::format`calls the format method of the superclass, on the same instance as `this`. However, `Formatter<ErrorOr<…>>` is not a superclass, and we don't really want multiple inheritance here.

Since we already do lots of allocations anyway, and printing ErrorOr's is not going to be on the hot path (at least for now), I'll happily let someone else do the optimizations. Let's merge it in as-is.

BenWiederhake commented:
Changes since last push:
- Rebased, and resolved trivial conflict in `System.{cpp,h}`
- Avoided an allocation in `Formatter<ErrorOr<>>`  if an error is being formatted (thanks @BertalanD!)

BenWiederhake commented:
Changes since last push:
* Rebased for convenience
* Undo the change in `Formatter<ErrorOr<…>>`: The suggested change cannot possibly work, as this Formatter is not a subclass of `Formatter<ErrorOr<…>>`. (CC @BertalanD)

IdanHo commented on Userland/Utilities/strace.cpp:
```diff
@@ -220,32 +221,27 @@ static void handle_sigint(int)
+static ErrorOr<ByteBuffer> copy_from_process(const void* source, size_t length)
 {
     auto buffer = ByteBuffer::create_uninitialized(length).value();
-    copy_from_process(source, buffer.bytes());
+    if (buffer.is_empty()) {
```
wrong check, we want to check if the optional returned is empty, not the buffer that we unconditionally unwrap.

IdanHo commented on Userland/Utilities/strace.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <LibC/sys/arch/i386/regs.h>
```
`benefits\!` -> `benefits!` in the commit message

IdanHo reviewed (commented):
Nice work, just a couple of comments

BenWiederhake commented:
Changes since last push:
- Fixed an error (Thanks IdanHo!)
- Fixed a silly typo in commit msg
- Rebased
